contracts/serverless-v2/Relay.sol

```diff
@@ -0,0 +1,482 @@
+
+    event JobRelayed(
+        uint256 indexed jobId,
+        bytes32 codehash,
+        bytes codeInputs,
+        uint256 userTimeout,    // in milliseconds
+        uint256 maxGasPrice,
+        uint256 usdcDeposit,
+        uint256 callbackDeposit,
+        address refundAccount,
+        uint256 startTime
+    );
+
+    event JobResponded(
+        uint256 indexed jobId,
+        bytes output,
+        uint256 totalTime,
+        uint256 errorCode,
+        bool success
+    );
+
+    event JobCancelled(uint256 indexed jobId);
+
+    error RelayInvalidUserTimeout();
+    error RelayJobNotExists();
+    error RelayOverallTimeoutOver();
+    error RelayInvalidJobOwner();
+    error RelayOverallTimeoutNotOver();
+    error RelayCallbackDepositTransferFailed();
+
+    //-------------------------------- internal functions start -------------------------------//
+
+    function _relayJob(
+        bytes32 _codehash,
+        bytes memory _codeInputs,
+        uint256 _userTimeout,   // in milliseconds
+        uint256 _maxGasPrice,
+        uint256 _callbackDeposit,
+        address _refundAccount,
+        address _jobOwner
+    ) internal {
+        if(_userTimeout <= GLOBAL_MIN_TIMEOUT || _userTimeout >= GLOBAL_MAX_TIMEOUT)
+            revert RelayInvalidUserTimeout();
+
+        if (jobCount + 1 == (block.chainid + 1) << 192)
+            jobCount = block.chainid << 192;
+
+        uint256 usdcDeposit = _userTimeout * EXECUTION_FEE_PER_MS + GATEWAY_FEE_PER_JOB;
+        jobs[++jobCount] = Job({
+            startTime: block.timestamp,
+            maxGasPrice: _maxGasPrice,
+            usdcDeposit: usdcDeposit,
+            callbackDeposit: _callbackDeposit,
+            jobOwner: _jobOwner
+        });
+
+        // deposit escrow amount(USDC)
+        TOKEN.safeTransferFrom(_jobOwner, address(this), usdcDeposit);
+
+        emit JobRelayed(jobCount, _codehash, _codeInputs, _userTimeout, _maxGasPrice, usdcDeposit, _callbackDeposit, _refundAccount, block.timestamp);
+    }
+
+    function _jobResponse(
+        bytes memory _signature,
+        uint256 _jobId,
+        bytes memory _output,
+        uint256 _totalTime,
+        uint8 _errorCode,
+        uint256 _signTimestamp
+    ) internal {
+        Job memory job = jobs[_jobId];
+        if(job.jobOwner == address(0))
+            revert RelayJobNotExists();
+
+        // check time case
+        if(block.timestamp > job.startTime + OVERALL_TIMEOUT)
+            revert RelayOverallTimeoutOver();
+
+        // signature check
+        address enclaveAddress = _verifyJobResponseSign(_signature, _jobId, _output, _totalTime, _errorCode, _signTimestamp);
+
+        address jobOwner = job.jobOwner;
+        uint256 callbackDeposit = job.callbackDeposit;
+        uint256 gatewayPayoutUsdc = _totalTime * EXECUTION_FEE_PER_MS + GATEWAY_FEE_PER_JOB;
+        uint256 jobOwnerPayoutUsdc = job.usdcDeposit - gatewayPayoutUsdc;
+        delete jobs[_jobId];
```

Comment:
job should be removed after the callback no?

Reply:
this is being deleted earlier to follow CEI pattern.

Reply:
sure, but doesn't the user contract need to be able to query job details if needed

Reply:
If those job details are required persistently, then we can skip deleting the job, or we can pass required job details in callback.

Reply:
how is the user's contract expected to handle the callback without any way of getting more context about what the callback is for